src/main/java/fi/aalto/cs/apluscourses/ui/base/TreeView.java

```diff
@@ -151,9 +151,12 @@ public void valueChanged(@NotNull TreeSelectionEvent e) {
 
     @Override
     public void mouseClicked(@NotNull MouseEvent e) {
-      if (e.getClickCount() == 2 && selectedItem != null) {
+      if (e.getClickCount() == 1 && e.isControlDown() && selectedItem != null) {
         ActionEvent actionEvent = new ActionEvent(this, ActionEvent.ACTION_PERFORMED, null);
-        nodeAppliedListeners.forEach(listener -> listener.actionPerformed(actionEvent));
+        nodeAppliedListeners.get("submitExercise").actionPerformed(actionEvent);
+      } else if (e.getClickCount() == 2 && selectedItem != null) {
+        ActionEvent actionEvent = new ActionEvent(this, ActionEvent.ACTION_PERFORMED, null);
+        nodeAppliedListeners.get("openSubmission").actionPerformed(actionEvent);
```

Comment:
TreeView should be a domain-ignorant class; it shouldn't deal with concepts such as "submission". It's a general purpose tree widget.

So if anything, the keys should be something like "doubleClick" and "ctrlClick", or maybe more preferably something abstract like "nodePrimaryAction", "nodeSecondaryAction".

I wouldn't, however, introduce any magic strings into the public interface of this class (they can be used internally, if you like). Instead, I would make a separate add\*Listener and remove\*Listener methods for different kind of events.

Reply:
I think separate add*Listener and remove*Listener methods for different kind of events sounds good, I'll do that

Reply:
Sounds good to me too, but in that case I would prefer two sets for primary and secondary listeners respectively, instead of a map (even if only used internally).

Reply:
Oops, Olli already suggested this in his later comment, but yes, I agree

Reply:
The javadocs for `add*Listener` could also mention when primary and secondary actions are triggered, since "primary" and "secondary" don't make it obvious